File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/minimum_window_substring/MinimumWindowSubstring.java
package main.java.com.jiang.leetcode100.minimum_window_substring;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Scanner;
import java.util.Set;

class Solution {
    public String minWindow(String s, String t) {

        int[] tmap = new int[256];
        for(int i = 0; i < t.length(); i++) tmap[t.charAt(i)]++;

        int[] smap = new int[256];

        int begin = 0, end = 0;
        int hasfound = 0;
        int min_size = Integer.MAX_VALUE;
        String result = "";
        while(end < s.length()) {
            smap[s.charAt(end)]++;
            if(smap[s.charAt(end)] <= tmap[s.charAt(end)]) hasfound++;
            if(hasfound == t.length() && tmap[s.charAt(end)] > 0) {
                while(smap[s.charAt(begin)] > tmap[s.charAt(begin)]) {
                    smap[s.charAt(begin)]--;
                    begin++;
                }
                if(end - begin + 1 < min_size) {
                    min_size = end - begin + 1;
                    result = s.substring(begin, end + 1);
                }
            }
            end++;
        }

        return result;
    }
}

public class MinimumWindowSubstring {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String s = scanner.nextLine();
        String t = scanner.nextLine();
        scanner.close();
        System.out.println(new Solution().minWindow(s, t));
    }
}

File: Leetcode100/leetcode100_python/MinimumWindowSubstring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        m = {}
        for ch in t:
            if ch in m:
                m[ch] += 1
            else:
                m[ch] = 1
        
        begin, end = 0, 0
        hasfound = 0
        result = ""
        while end < len(s):
            if s[end] in m:
                m[s[end]] -= 1
            else:
                m[s[end]] = -1
            if m[s[end]] >= 0:
                hasfound += 1
            if hasfound == len(t):
                while m[s[begin]] < 0:
                    m[s[begin]] += 1
                    begin += 1
                if not result or end - begin + 1 < len(result):
                    result = s[begin : end + 1]
                m[s[begin]] += 1
                begin += 1
                hasfound -= 1
            end += 1
        return result

if __name__ == "__main__":
    s = input()
    t = input()
    solution = Solution()
    print(solution.minWindow(s, t))